### PR TITLE
Report failing and should_fail tests as XFAIL, instead of OK

### PR DESCRIPTION
**Pull request by falconindy**
This makes it clear that there's "passing" tests which actually require
bugfixes in the project, e.g. instead of:

  OK:       329
  FAIL:       1
  SKIP:       0
  TIMEOUT:    0

Users now see:

  OK:       323
  FAIL:       1
  XFAIL:      6
  SKIP:       0
  TIMEOUT:    0

### CODE REVIEW

**Comment by QuLogic** [comment]
Going by pytest nomenclature, isn't this XPASS?

**Comment by jpakkane** [comment]
> This makes it clear that there's "passing" tests which actually require bugfixes in the project,

This is not actually true. Using expected failures to mark bugs is an antipattern and you should not do that. Tests that should fail are for things like trying to run an executable with known invalid input. It _must_ return an error in this case rather than silently accept it.

But ignoring that the UX for the word "XFAIL" is confusing, because it shows you something like "XFAIL: 6" and you always have to spend the extra mental effort to realize that it does not mean failure but success.

As an example there is a certain test framework runner that shall not be named that prints "ERRORS!" (meaning zeror errors) intermixed between other text and it is _so_ annoying to have that there.

Having a case where everything went as expected prints out FAIL in any way does not seem nice. Thus XPASS would make more sense, but its meaning as a word is not as clear. In any case it should be printed just below OK so all passing categories are printed first, then failing ones and then the rest.

**Comment by falconindy** [comment]
> Using expected failures to mark bugs is an antipattern and you should not do that. Tests that should fail are for things like trying to run an executable with known invalid input. It must return an error in this case rather than silently accept it.

I'd hardly call this an anti-pattern. There's an entire [development strategy](https://en.wikipedia.org/wiki/Test-driven_development) based on this principle.

I'm unclear on the state of this PR. Would you like to see the output reordered? XFAIL changed to XPASS? Following your suggestion that "ok" things be printed first, should we also move "SKIP" higher up? Autotools uses this ordering:

```
# TOTAL: 522
# PASS:  516
# SKIP:  0
# XFAIL: 6
# FAIL:  0
# XPASS: 0
# ERROR: 0
```

**Comment by falconindy** [comment]
This was also requested in #3296.

Happy to rework if to include both XPASS and XFAIL, but some direction is needed from the maintainers.

**Comment by jpakkane** [comment]
As a feature this is nice, it's just that the XPASS and XFAIL are logically backwards and you need to know that the X stands _either_ for "unexpected" _or_ "expected". Could it be possible to come up with a better, unambiguous nomenclature here?

**Comment by falconindy** [comment]
I mean, we could eschew any notion of short hand and just spell it out.

Pass, fail, expected fail, unexpected pass, etc.

File: mesonbuild/mtest.py [hunk 1]
```diff
@@ -499,11 +508,14 @@ def print_stats(self, numlen, tests, name, result, i):
...
-TIMEOUT: %4d
-''' % (self.success_count, self.fail_count, self.skip_count, self.timeout_count)
+Ok:                 %4d
+Expected Fail:      %4d
+Fail:               %4d
+Unexpected Pass:    %4d
+Skipped:            %4d
+Timeout:            %4d
+''' % (self.success_count, self.expectedfail_count, self.fail_count,
+       self.unexpectedpass_count, self.skip_count, self.timeout_count)
```
Comment (ignatenkobrain) [inline mesonbuild/mtest.py]: __[Flake8]__

[E122] continuation line missing indentation or outdented

**Comment by falconindy** [comment]
So this now spells out EXPECTEDFAIL and UNEXPECTEDPASS.

**Comment by jpakkane** [comment]
Works for me. Thanks.

**Comment by heftig** [comment]
I would have called it "Pass:" or "Unexpected Ok:" for symmetry.

**Pull request merged**